rust-runtime/aws-smithy-http-server/src/plugin/check_health.rs

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::future::Ready;
+
+use futures_util::Future;
+use http::StatusCode;
+use hyper::{Body, Request, Response};
+use tower::Layer;
+use tower::Service;
+
+use crate::body;
+use crate::body::BoxBody;
+
+use super::Either;
+
+#[derive(Clone)]
+pub struct CheckHealthLayer<H> {
+    ping_handler: H,
+}
+
+impl<H> CheckHealthLayer<H> {
+    pub fn new(ping_handler: H) -> Self {
+        CheckHealthLayer { ping_handler }
+    }
+}
+
+pub type DefaultHandler<E> = fn(Request<Body>) -> Ready<Result<Response<BoxBody>, E>>;
+
+impl CheckHealthLayer<()> {
+    pub fn with_default_handler<E>() -> CheckHealthLayer<DefaultHandler<E>> {
+        CheckHealthLayer::new(default_ping_handler)
+    }
+}
+
+impl<S, H: Clone> Layer<S> for CheckHealthLayer<H> {
+    type Service = CheckHealthService<H, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CheckHealthService {
+            inner,
+            layer: self.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct CheckHealthService<H, S> {
+    inner: S,
+    layer: CheckHealthLayer<H>,
+}
+
+impl<H, HandlerFuture, S> Service<Request<Body>> for CheckHealthService<H, S>
+where
+    S: Service<Request<Body>, Response = Response<BoxBody>> + Clone,
+    S::Future: std::marker::Send + 'static,
+    HandlerFuture: Future<Output = Result<Response<BoxBody>, S::Error>>,
+    H: Fn(Request<Body>) -> HandlerFuture,
+{
+    type Response = S::Response;
+
+    type Error = S::Error;
+
+    type Future = Either<HandlerFuture, S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
```

Comment:
This `Service` is kinda a "router" style branch. The `poll_ready` is actually too restrictive for the case where the `http::Request` directs us to the ping handler.

_Maybe_ what we want here is `Poll::Ready(Ok())` and then in the handler we use `ServiceExt::oneshot` on the inner service.

Reply:
@LucioFranco What's the guidance here? If there's a service which contains multiple services, each of which are called depending on control flow controlled by the `Request`, then what should we do:

a) Check that all inner services are `Poll::Ready`?
b) Always return `Poll::Ready` and use `oneshot` in the `call`?

Reply:
So I think you can look at balance/steer in the tower repo to see how they handle this stuff. Though in this case I feel like since this is server side and generally on server side you don't really use back pressure since the only real mechanism for back pressure is load shedding (and for h2 not reading from the body which stops the window from moving). So that being said I feel like here you can do what @hlbarber said and return read and then oneshot. I would maybe even review the middleware that you expect to be below this in the stack and see if they do anything with poll_ready it might be the case that they don't do anything and its safe to just call directly into call.

Reply:
I went for `(b)`, pushed in bdbed1b012aa191884021398850deefb5a7e63f0.